src/ImgixTag.js

```diff
@@ -0,0 +1,155 @@
+var util = require('./util.js'),
+    targetWidths = require('./targetWidths.js');
+
+var ImgixTag = (function() {
+  function ImgixTag(el) {
+    this.el = el;
+
+    if (!this.el) {
+      throw new Error('ImgixTag must be passed a DOM element.');
+    }
+
+    if (this.el.hasAttribute('ix-initialized')) {
+      return;
+    }
+
+    this.ixPathVal = el.getAttribute('ix-path');
+    this.ixParamsVal = el.getAttribute('ix-params');
+    this.ixSrcVal = el.getAttribute('ix-src');
+
+    if (this.ixPathVal && !imgix.config.host) {
+      throw new Error('You must set a value for `imgix.config.host` to use `ix-path` and `ix-params`.');
+    }
+
+    this.baseParams = this._extractBaseParams();
+    this.baseUrl = this._buildBaseUrl();
+    this.baseUrlWithoutQuery = this.baseUrl.split('?')[0];
+
+    this.el.setAttribute('sizes', this.sizes());
+    this.el.setAttribute('srcset', this.srcset());
+
+    if (this.el.nodeName == 'IMG') {
+      this.el.setAttribute('src', this.src());
+    }
+
+    this.el.setAttribute('ix-initialized', 'ix-initialized');
+  }
+
+  ImgixTag.prototype._extractBaseParams = function() {
+    var params;
+
+    if (this.ixPathVal) {
+      params = JSON.parse(this.ixParamsVal) || {};
+    } else {
+      // If the user used `ix-src`, we have to extract the base params
+      // from that string URL.
+      var lastQuestion = this.ixSrcVal.lastIndexOf('?'),
+          paramString = this.ixSrcVal.substr(lastQuestion + 1),
+          splitParams = paramString.split('&');
+
+      params = {};
+      for (var i = 0, splitParam; i < splitParams.length; i++) {
+        splitParam = splitParams[i].split('=');
+
+        params[splitParam[0]] = splitParam[1];
+      }
+    }
+
+    // Encode any passed Base64 variant params
+    for (var key in params) {
+      if (key.substr(-2) === '64') {
+        params[key] = util.encode64(params[key]);
+      }
+    }
+
+    return params;
+  };
+
+  ImgixTag.prototype._buildBaseUrl = function() {
+    if (this.ixSrcVal) {
+      return this.ixSrcVal;
+    } else {
+      var path = this.ixPathVal,
+          protocol = 'http';
+      if (imgix.config.useHttps) {
+        protocol += 's';
+      }
+
+      var url = protocol + '://' + imgix.config.host,
+          hostEndsWithSlash = imgix.config.host.substr(-1) === '/',
+          pathStartsWithSlash = path[0] === '/'
+
+      // Make sure we don't end up with 2 or 0 slashes between
+      // the host and path portions of the generated URL
+      if (hostEndsWithSlash && pathStartsWithSlash) {
+        url += path.substr(1);
+      } else if (!hostEndsWithSlash && !pathStartsWithSlash) {
+        url += '/' + path;
+      } else {
+        url += path;
+      }
+
+      url += '?'
+      var params = [],
+          param;
+      for (var key in this.baseParams) {
+        param = this.baseParams[key];
+        params.push(encodeURIComponent(key) + '=' + encodeURIComponent(param));
```

Comment:
It's a _very, very_ edgecasey sort of thing, but one we've been moving all the libs toward handling. Theoretically, someone could accept user input for keys.